PERF: float hash slow in py3

chris-b1 opened the pull request:
closes #13166, closes #13335

Using exactly the approach suggested by @ruoyu0088

significant changes in asv below

```
     before     after       ratio
-     8.88s    78.12ms      0.01  indexing.float_loc.time_float_loc
-    13.11s    78.12ms      0.01  groupby.groupby_float32.time_groupby_sum
```

Factor of 10 smaller benches

```
     before     after       ratio
-  171.88ms    43.29ms      0.25  indexing.float_loc.time_float_loc
-     1.42s    11.23ms      0.01  groupby.groupby_float32.time_groupby_sum
```

chris-b1 committed:
PERF: float hash slow in py3

jreback commented:
wow those were slow before, can you make 10x smaller in the asv (so the slow ones don't take as much time). you should still have the same ratio.

jreback commented on pandas/src/klib/khash_python.h:
```diff
@@ -2,9 +2,10 @@
-
-#define kh_float64_hash_func _Py_HashDouble
+inline khint64_t asint64(double key) {
+  return *(khint64_t *)(&key);
```
add a comment (and a link to the python source code) of why we are not using ``_Py_HashDouble`` here (and this issue number as well)

chris-b1 committed:
smaller benches, explanatory comment

chris-b1 commented:
@jreback - I've updated this.  It took a while to figure out, but the issue here isn't actually the time of the python hash function, it's that in py3 it returns a `Py_ssize_t` (was `long`), and our khash keys are 32 bit, so the truncation is causing the collisions.  I updated the asv at the top with the smaller benches - because it's a collision issue the ratios aren't both the same, but still show the issue.

```python
In [50]: a = np.arange(1000000)
    ...: ind = pd.Float64Index(a * 4.8000000418824129e-08)

In [51]: hashes = np.array([hash(x) for x in ind])

In [52]: len(hashes), len(pd.unique(hashes))
Out[52]: (1000000, 1000000)

In [53]: truncated = hashes.view('int32')[::2]

In [54]: len(truncated), len(pd.unique(truncated))
Out[54]: (1000000, 524288)
```

jreback commented:
ty sir!
